components/director/pkg/operation/update_operation_handler.go

```diff
@@ -0,0 +1,156 @@
+	return &updateOperationHandler{
+		transact:             transact,
+		resourceUpdaterFuncs: resourceUpdaterFuncs,
+		resourceDeleterFuncs: resourceDeleterFuncs,
+	}
+}
+
+// ServeHTTP handles the Operations API requests
+func (h *updateOperationHandler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	ctx := request.Context()
+
+	if request.Method != http.MethodPut {
+		apperrors.WriteAppError(ctx, writer, apperrors.NewInternalError("Method not allowed"), http.StatusMethodNotAllowed)
+		return
+	}
+
+	bytes, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		apperrors.WriteAppError(ctx, writer, apperrors.NewInternalError("Unable to read request body"), http.StatusInternalServerError)
+		return
+	}
+	defer func() {
+		err := request.Body.Close()
+		if err != nil {
+			log.C(ctx).WithError(err).Error("Failed to close request body")
+		}
+	}()
+
+	var operation OperationRequest
+	if err := json.Unmarshal(bytes, &operation); err != nil {
+		apperrors.WriteAppError(ctx, writer, apperrors.NewInternalError("Unable to decode body to JSON"), http.StatusBadRequest)
+		return
+	}
+
+	if err := validation.ValidateStruct(&operation,
+		validation.Field(&operation.ResourceID, is.UUID),
+		validation.Field(&operation.OperationType, validation.Required, validation.In(graphql.OperationTypeCreate, graphql.OperationTypeUpdate, graphql.OperationTypeDelete)),
+		validation.Field(&operation.ResourceType, validation.Required, validation.In(resource.Application))); err != nil {
+		apperrors.WriteAppError(ctx, writer, apperrors.NewInvalidDataError("Invalid operation properties: %s", err), http.StatusBadRequest)
+		return
+	}
+
+	tx, err := h.transact.Begin()
+	if err != nil {
+		log.C(ctx).WithError(err).Errorf("An error occurred while opening db transaction: %s", err.Error())
+		apperrors.WriteAppError(ctx, writer, apperrors.NewInternalError("Unable to establish connection with database"), http.StatusInternalServerError)
+		return
+	}
+	defer h.transact.RollbackUnlessCommitted(ctx, tx)
+
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	resourceUpdaterFunc := h.resourceUpdaterFuncs[operation.ResourceType]
+	switch operation.OperationType {
+	case graphql.OperationTypeCreate:
+		fallthrough
+	case graphql.OperationTypeUpdate:
+		if err := resourceUpdaterFunc(ctx, operation.ResourceID, true, getError(operation.Error)); err != nil {
+			log.C(ctx).WithError(err).Errorf("While updating resource %s with id %s", operation.ResourceType, operation.ResourceID)
+			apperrors.WriteAppError(ctx, writer, apperrors.NewInternalError("Unable to update resource %s with id %s", operation.ResourceType, operation.ResourceID), http.StatusInternalServerError)
+			return
+		}
+	case graphql.OperationTypeDelete:
+		resourceDeleterFunc := h.resourceDeleterFuncs[operation.ResourceType]
+		if operation.Error != "" {
+			if err := resourceUpdaterFunc(ctx, operation.ResourceID, true, getError(operation.Error)); err != nil {
```

Comment:
The API seems to fail deep in the storage layer whenever there's an error to persist - the raw SQL error returns something of the kind: `Token "reconciliation is invalid` (the original error message to persist was `reconciliation timeout reached`